From ec16338243bb4554435e96084ad95bad25639b5a Mon Sep 17 00:00:00 2001
From: Shammah Chancellor <shammah.chancellor@notvalid.com>
Date: Sun, 18 Oct 2026 01:33:29 -0700
Subject: [PATCH] miner: overhaul BlockAssembler to O(1) Merklix overlay
 assembly

Update BlockAssembler::CreateNewBlock to assemble block templates from
the speculative Merklix overlay.

diff --git a/src/test/util/mining.cpp b/src/test/util/mining.cpp
index 44da1beba8c002e4461bedd5824ab23c32ebb366..b60095383a8470245278a6dc02ca0d32570f90a0 100644
--- a/src/test/util/mining.cpp
+++ b/src/test/util/mining.cpp
@@ -61,7 +61,9 @@ PrepareBlock(const Config &config, const NodeContext &node,
 
     LOCK(cs_main);
     block->nTime = Assert(node.chainman)->ActiveTip()->GetMedianTimePast() + 1;
-    block->hashMerkleRoot = BlockMerkleRoot(*block);
+    if (block->hashMerkleRoot.IsNull()) {
+        block->hashMerkleRoot = BlockMerkleRoot(*block);
+    }
 
     return block;
 }
